Small optimizations, and code refactor

simonmeulenbeek opened the pull request:
Hello RO maintainers!

I was trying to figure out how RSSTimeFormatter does some things, but while reading the code I noticed some flaws that I'd thought I'd fix. 

**Important:**
This changes the functionality of `PrintTime(double time, int valuesOfInterest, bool explicitPositive)`. This fixes issue #12, and flips the behaviour of `valuesOfInterest`. 
I.e.:
```
PrintTime("1y 10d 8h 5m 12s", 0, false) == "1y"
PrintTime("1y 10d 8h 5m 12s", 1, false) == "1y, 10d"
PrintTime("1y 10d 8h 5m 12s", 2, false) == "1y, 10d, 8h"
```

**What changed?**
I'll add some code comments too but an example is this bit that returns in many functions:
```
DateTime epoch = GetEpoch();
DateTime target = epoch.AddSeconds(time);
TimeSpan span = target - epoch;
````
some issues with this: 
* `epoch.AddSeconds()` returns a new instance, so the class-level epoch-object will not be changed. It is therefore unnecessary to temporarily create the epoch object in each function. 
* The resulting `TimeSpan span` will always be of length `time`. So this code should be replaced by `TimeSpan.FromSeconds(time);`

I've also refactored some common functions to it's own code. This might be a little bit verbose to your liking, but IMO makes it easier to read and thus maintain.

simonmeulenbeek commented on RSSTimeFormatter/DTReplacer.cs:
```diff
@@ -9,57 +9,58 @@ public class DTReplacer : MonoBehaviour
 		public static string GetUniqueStringFromUniqueNode(string name, string node)
 		{
 			var configs = GameDatabase.Instance.GetConfigs(node);
-			if (configs.Length > 1) {
-				Debug.LogError(
-					"Multiple `" + node + "` configurations, falling back to default");
-			}
-			else if (configs.Length == 1) {
-				ConfigNode config = configs[0].config;
-				var formats = config.GetValues(name);
-				if (formats.Length > 1) {
-					Debug.LogError(
-						"`" + node + "` configuration has multiple `" + name + "` entries, falling back to default");
-				}
-				else if (formats.Length == 1) {
-					return formats[0];
-				}
-			}
-			return null;
```
Because we're only interested in if the config and formats length are 1. We can reduce the number of if-checks, if we reverse the order and just check if length == 1 directly.

RCrockford commented on RSSTimeFormatter/RealDateTimeFormatter.cs:
```diff
@@ -6,257 +6,287 @@
+            time = Math.Abs(time);
+            TimeSpan span = TimeSpan.FromSeconds(time);
+            int years = (int)(span.TotalDays / 365);
+            return string.Format("{0}{1}{2}{3}{4}{5}"
```
If `valuesOfInterest` is 1 and you have a time of less than a day, this will not print anything. I don't think this is intended behaviour?

simonmeulenbeek replied:
Well, I don't know. Depends on how the original KSP interface is designed I guess.
If `valuesOfInterest` is supposed to set the significance of the answer, this would be correct. If you're only interested in the amount of years, a length of less than a year should be either '0' or maybe '<1' .
If `valuesOfInterest` is supposed to set the amount of fields returned it is indeed wrong. If the latter is the case, the fix would be easy, but it depends on the intended design.

RCrockford replied:
There is an issue that you don't get anything if the number of years (/months/etc) is less than one. `PrintTime(1, 1, false)` yields an empty string for example, not "0".

jwvanderbeck replied:
The original KSP code (or at least a version made from decompile) is here for reference: https://github.com/KSP-RO/RSSTimeFormatter/blob/master/RSSTimeFormatter/DefaultDateTimeFormatter.cs#L82-L109

simonmeulenbeek replied:
I've changed `PrintTime()` to be like the original KSP code. I.e. it will give `valuesOfInterest` amount of values from the first non-zero value.
e.g.:
```
PrintTime("0y 1d 2h 3m 4s", 0, false) = "1d"
PrintTime("0y 0d 2h 3m 4s", 0, false) = "2h"
PrintTime("0y 1d 2h 3m 4s", 1, false) = "1d, 2h"
PrintTime("1y 0d 0h 3m 4s", 1, false) = "1y, 0d"
PrintTime("1y 0d 0h 3m 4s", 2, false) = "1y, 0d, 0h"
```

siimav commented:
I haven't really taken a in-dept look yet but a bunch of things stood out right away:
* You're using java-style braces
* For properties, use https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/properties#expression-body-definitions
* A bunch of missing empty lines before methods
* Method names need to be pascal case
* Mixed tabs and spaces. Looks like it was tabs-only previously? If so then your style should be consistent with was used used in the file previously. I might convert all the tabs to spaces myself at a later time.
* String interpolation is way better than string.format

siimav commented:
Btw, C# DateTime is a struct and won't cause heap allocations

simonmeulenbeek commented:
@siimav 
I've asked in the Discord if there was a particular style guide, but couldn't find any. I think VStudio reformatted a lot of things on import, because there were no newlines between functions. It's no problem to fix the formatting on my side, if I know what it should be :smile: 


> You're using java-style braces
You've caught the Java developer :blush: 

> For properties, use https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/properties#expression-body-definitions
What do you mean? Can you point me to where I'm using these? I didn't add the Day/year/etc properties.

> A bunch of missing empty lines before methods
Done
> Method names need to be pascal case
Done
> Mixed tabs and spaces. Looks like it was tabs-only previously? If so then your style should be 
consistent with was used used in the file previously. I might convert all the tabs to spaces myself at a later time.
I've changed it to all-tabs for now. 

>String interpolation is way better than string.format
For readability mostly yes. I've changed most string.formats to interpolation, except for the ones which have boolean operations. Also in the case of some functions which use a seperator I've found it more readable to use string.format.

jwvanderbeck commented:
> @siimav
> I've asked in the Discord if there was a particular style guide, but couldn't find any. I think VStudio reformatted a lot of things on import, because there were no newlines between functions. It's no problem to fix the formatting on my side, if I know what it should be 😄

In the absence of direction, the style is what is present :)

Generally though, the accepted C# community standard (NOT Unity's C# standard lol)

> > String interpolation is way better than string.format
> 
> For readability mostly yes. I've changed most string.formats to interpolation, except for the ones which have boolean operations. Also in the case of some functions which use a seperator I've found it more readable to use string.format.

Yeah, we should definitely be moving towards the new style `$""` string interpolation whenever possible which wasn't an available language feature when this was originally written.  It is so much cleaner.

Give me some time to review this today.  Thanks for the PR though.  I was going to sit down and address that issue at some point but kept never getting to it.

jwvanderbeck commented on RSSTimeFormatter/RealDateTimeFormatter.cs:
```diff
@@ -100,131 +82,192 @@ public string PrintTimeCompact(double time, bool explicitPositive)
 				return InvalidTimeStr(time);
-			if (time < 0 && useAbs)
-				time = Math.Abs(time);
+			time = useAbs ? Math.Abs(time) : time;
```
I really dislike the use of this style conditional.  I know, its personal taste, but I prefer clear verbose `if` statements

simonmeulenbeek replied:
>  I prefer clear verbose if statements
I do to! I've changed it to:
`if (useAbs) time = Math.Abs(time);`

the `time < 0` is unnecessary because Math.Abs() effectively doesn't do anything if time > 0.

jwvanderbeck replied:
Fair enough but could still do it with an if :)

siimav replied:
That is definitely down to your personal taste. I do agree that the inline conditionals can get messy in some cases but this one isn't the case.

jwvanderbeck commented:
I need more time to go through this, but I feel like there are too many changes here that aren't required.  You've made fixes and some functional changes which awesome are welcome.  But you've also taken it upon yourself to just refactor/cleanup code that makes no functional difference. I would much rather the PR stick to just the functional changes.

simonmeulenbeek commented:
>  I need more time to go through this, but I feel like there are too many changes here that aren't required. You've made fixes and some functional changes which awesome are welcome. But you've also taken it upon yourself to just refactor/cleanup code that makes no functional difference. I would much rather the PR stick to just the functional changes.

Hmm yes, this PR has grown beyond it's original intended scope. Do you want me to cancel this PR and put in a new one with just the PrintTime() fix? There are no other functional changes.

jwvanderbeck commented:
> > I need more time to go through this, but I feel like there are too many changes here that aren't required. You've made fixes and some functional changes which awesome are welcome. But you've also taken it upon yourself to just refactor/cleanup code that makes no functional difference. I would much rather the PR stick to just the functional changes.
> 
> Hmm yes, this PR has grown beyond it's original intended scope. Do you want me to cancel this PR and put in a new one with just the PrintTime() fix? There are no other functional changes.

Yeah that would be appreciated, or you can just update this one. Whichever is easier for you.
